src/fragmentarium/ui/fragment/TeiExport.js

```diff
@@ -0,0 +1,138 @@
+export function teiExport(fragment) {
+  var res = ''
+  const header = getHeader(fragment)
+  const body = getBody(fragment)
+  const end = getEnd()
+
+  res += header
+  res += body
+  res += end
+
+  console.log(fragment)
+
+  return prettifyXml(res)
+}
+
+function getHeader(fragment) {
+  var res = ''
+  const start =
+    '<?xml version="1.0" encoding="UTF-8"?><TEI xmlns="http://www.tei-c.org/ns/1.0"><teiHeader><fileDesc>'
+  const title_stmt =
+    '<titleStmt><title>' + fragment.number + '</title></titleStmt>'
+  const pub_stmt =
+    '<publicationStmt><p>' +
+    fragment.museum.name +
+    '</p><p>' +
+    fragment.collection +
+    ' Collection</p></publicationStmt>'
+
+  var source_desc = '<sourceDesc><p>' + fragment.description + '</p>'
+
+  if (fragment.publication)
+    source_desc += '<p>Publication: ' + fragment.publication + '</p>'
+  // if (fragment.references.length>0)  source_desc += getBibiliography(fragment);
+
+  const end = '</sourceDesc></fileDesc></teiHeader><text><body>'
+
+  return res + start + title_stmt + pub_stmt + source_desc + end
+}
+
+// function getBibiliography(fragment){
+// var res = '<biblFull>'
+
+// Object.values(fragment.references).forEach(val => {
+//  res+=addSingleReference(val);
+// })
+
+// res+= '</biblFull>'
+// return res
+// }
+
+// function addSingleReference(ref){
+// var res = '<titleStmt>';
+// res += '<title>'+ ref.document.cslData.title + '</title>';
+
+// var short_author = ref.document.cslData['container-title-short']
+// var author = ref.document.cslData.author
+
+// if(author || short_author) {
+//   res += '<author>'
+//     if(short_author)res += short_author
+//     else res += author[0].given + " "+author[0].family
+//   res += '</author>'
+// }
+
+// res += '</titleStmt>'
+// return res
+// }
+
+function getEnd() {
+  return '</body></text></TEI>'
+}
+
+function getBody(fragment) {
+  return fragment.text.allLines.filter(isNotEmpty()).map(getParagraph).join('')
+}
+
+function isNotEmpty() {
+  return (line) => line.type !== 'EmptyLine'
+}
+
+function getParagraph(line) {
+  var res = ''
+
+  if (line.type === 'TextLine') {
+    res += '<p n="' + line.lineNumber.number + '">'
+    // res += line.prefix  // line numbers needed in text?
+    res += getLineContent(line)
+    res += '</p>'
+  } else {
+    res += '<p>'
+    res += line.prefix
+    res += getLineContent(line)
+    res += '</p>'
+  }
+
+  return res
+}
+
+function getLineContent(line) {
+  var res = ''
+  for (var j = 0; j < line.content.length; j++) {
```

Comment:
Do not use generic one letter names (i, j, k, x, ...). I.e use `index` for indices.

Reply:
reject. makes code more unreadable and i as loop counter is widespread convention.

Reply:
I disgagree that shortening automatically improves readability, but in a simple loop, this should not be an issue, and single letter names are used in [generics](https://www.typescriptlang.org/docs/handbook/generics.html). Although `i` should be used. If the function has multiple loops it needs refactoring. In the latest code `i` is also used in other functions, where better naming would be required or it must be kept strictly within the loop body (e.g. pass current and previous line).

The for-loop itself is not very readable. In addition to keeping track of the index and the result variable, the reader needs to figure out the purpose of the if. In the case above `map` + `join` would be more readable. It clearly communicates the intention of the code and gets rid of the intermediate variables. 

I do no net see many uses for the plain for-loop. [for..of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of), `map`, `filter` and `forEach`, remove the boiler plate of accessing the element at index and can communicate the intent better. Arrays methods also have extra arguments for `index` and `array`.

Reply:
in the latest version, i is changed to previous_line. I disagree on using map, filter since it is 10 times slower (https://dev.to/henryjw/array-map-much-slower-than-for-loop-57if) than regular for loop and increases memory consumption due to copying arrays. We discussed performance lately and now you propose a construction that will significantly slow down the code, that does not appear reasonable to me.

Reply:
Is the page too slow to use? Did you profile it? On different browsers?

The code should be made good first, only then fast. If using a more readable construction is not "too slow", there is need to do premature optimization. I raised the issue with string catenation because I was iterested in the effect it might have. It was investigated and no action deemed necessary, even if it seemed to use lot of memory compared to produced the XML size.

If it indeed is too slow then there might be other optimizations consider. See the comment about lodash.

Reply:
I did quick tests with `performance.now()` on Edge (Other browsers might give different results.) and there appears to no performance difference. There were some outliers with long execution time, but the test fragment mostly took 1,6-2 s and the largest 7-9 s. 

10 times slower is only valid in the context of that micro-benchmark from over a year ago. Browsers might have been optimized after that, or the lost performance is won by reducing boilerplate code for managing the loops. These benchmarks can be useful to find possible optimizations after bottlenecks have been identified. Still, to avoid sub-optimization, the whole context must be taken into account when deciding if/what to optimize.

In this case, it appears that the code was made less maintainable for no benefit at all. Please abstain from making premature optimizations in the future.

However, the execution time of several seconds would call for optimizations. It appears that not prettifying could shave about 2 seconds from the largest fragment. Specific elements are always at the same level. Maybe the indentation could be added together with the element. Alternatively, the generation of the XML could be delayed until the user clicks the downloads button. This would remove the performance loss from the users who are not interested in the XML.

Reply:
Can you send/commit the version you tested against the current version somewhere?

Reply:
Got unit wrong in the last message 🤦 , it's hard to concentrate when the kid is at home. Should be milliseconds not seconds. Does not change the point though, expect that optimization should not be necessary.

Here's the commit https://github.com/ElectronicBabylonianLiterature/ebl-frontend/commit/66099b52a849c328da4e1f6b2527075cd8c4f320

Reply:
Ok, that appears to be more realistic ;) It does change the point of creating the export only on clicking the button though. Thanks for the commit i'll also do a performance test when i find the time.